skimage/_shared/tests/test_utils.py

```diff
@@ -347,3 +349,318 @@ def test_deprecate_func():
     )
     assert record[0].lineno == expected_lineno
     assert record[0].filename == __file__
+
+
+@deprecate_parameter("old1", start_version="0.10", stop_version="0.12")
+@deprecate_parameter("old0", start_version="0.10", stop_version="0.12")
+def _func_deprecated_params(arg0, old0=DEPRECATED, old1=DEPRECATED, arg1=None):
+    """Expected docstring.
+
+    Parameters
+    ----------
+    arg0 : int
+        First unchanged parameter.
+    arg1 : int, optional
+        Second unchanged parameter.
+    """
+    return arg0, old0, old1, arg1
+
+
+@deprecate_parameter("old1", new_name="new0", start_version="0.10", stop_version="0.12")
+@deprecate_parameter("old0", new_name="new1", start_version="0.10", stop_version="0.12")
+def _func_replace_params(
+    arg0, old0=DEPRECATED, old1=DEPRECATED, new0=None, new1=None, arg1=None
+):
+    """Expected docstring.
+
+    Parameters
+    ----------
+    arg0 : int
+        First unchanged parameter.
+    new0 : int, optional
+        First new parameter.
+
+        .. versionadded:: 0.10
+    new1 : int, optional
+        Second new parameter.
+
+        .. versionadded:: 0.10
+    arg1 : int, optional
+        Second unchanged parameter.
```

Comment:
Hmm... Since `arg1` is added in version 0.10, I understand that no change happens to it (at version 0.12, in the context of this mock function), but I guess its description shouldn't be exactly parallel to `arg0`'s, which was 'always' there, i.e., wasn't added in version 0.10.
```suggestion
        Parameter added (not replacing any other parameter).
```

Reply:
I am not following. In this test function `arg1` is a parameter that is unaffected by the decorators. At least to the extend that it was there before. Adding the decorators only affects it in the sense that it's no longer reachable as a positional parameter because that would require passing deprecated parameters before, e.g.:

```python
_func_deprecated_params(0, 1, 2, 3)
```
3 is passed to `arg1` but 1, 2 are triggering deprecation warnings. 

This is a quirk of deprecating positional arguments. Positional args can only be reached by triggering a deprecation warning if there are deprecated positional args in front. But that's also very useful because at the end of the deprecation we can switch the remaining parameters to keyword-only args!

Reply:
Oops, sorry, because of the blank line before
```py
        .. versionadded:: 0.10
```
my eyes associated it with what's below (`arg1 : int, optional` instead of above (`new1 : int, optional`).

> But that's also very useful because at the end of the deprecation we can switch the remaining parameters to keyword-only args!

Nice! So, at the end of the deprecation cycle, we'll have the following function signature
```py
def _func_replace_params(
    arg0, new0=None, new1=None, *, arg1=None
):
```
? Or even
```py
def _func_replace_params(
    arg0, *, new0=None, new1=None, arg1=None
):
```
? Maybe I'll find the answer as I finish reviewing... :pray:

Reply:
Correct. After the deprecation cycle we can be sure that users always got a warning if they used any parameter following `old0` as a positional one. So we can move over to `def _func_replace_params(arg0, *, new0=None, new1=None, arg1=None)` :tada:.

Thinking about this, there is some case that is somewhat tricky to deprecate with a grace period where both APIs work at the same time. When changing
```python
def foo(a, b, c):  # old API

# ???

def foo(a, c):  # new API
```
`foo(1, 2)` throws a warning because `b` is deprecated. But if users still want to use `c` while avoiding the warning and being future proof, they have to use `foo(1, c=2)`. So `b` needs a default parameter in the meantime, which also makes `c` need a default parameter, at least temporarily. Do you follow? I find this hard to explain. :sweat_smile:

Reply:
Oh, sorry, maybe I'm getting only now what you really(?) meant: The utilities you're providing with this PR will let us handle the deprecation (any deprecation) on a 'formal' ('external') level but, in cases such as what you're describing, we'll have to adapt the actual code (inside the function) to handle all cases during the deprecation period: if function call is `foo(1, 2, 3)` elif `foo(1, 2)` elif `foo(1, c=2)` else... error? So, yes, it will be more work that just decorating the function with the new `deprecate_parameter` helper. Is that what you meant?

Reply:
Yes, your second guess is correct. Though, I argue that in cases I'm describing, it's very difficult to do a proper deprecation path because a new and old positional API can't be valid at the same time.

Reply:
Right...